mdsuite/transformations/map_molecules.py

```diff
@@ -119,286 +118,175 @@ def _run_dependency_check(self):
+            batch_generator, args=batch_generator_args, output_signature=type_spec
+        )
+        data_set = data_set.prefetch(tf.data.experimental.AUTOTUNE)
 
-        Parameters
-        ----------
-        species : list
-                List of species to be loaded.
-        Returns
-        -------
-        mass_array : list
-                A list of masses.
-        """
-        mass_array = []
-        for item in species:
-            mass_array.append(self.experiment.species[item]["mass"])
+        log.info(f"Mapping molecule graphs onto trajectory for {molecule_name}")
+        for i, batch in tqdm(enumerate(data_set), ncols=70, total=self.n_batches):
+            batch_size = batch[b"data_size"]
 
-        return mass_array
+            trajectory = np.zeros(
+                shape=(
+                    molecular_graph.n_molecules,
+                    batch_size,
+                    3,
+                )
+            )
 
-    def _map_molecules(self):
-        """
-        Map the molecules and save the data in the database.
+            for t, molecule in enumerate(molecular_graph.molecular_groups):
+                # Load species molecule-specific particles into a separate dict
+                # and apply their respective scaling factor.
+                molecule_trajectory = np.zeros((batch_size, 3))
+                for item in molecular_graph.molecular_groups[molecule]:
+                    batch_reference = str.encode(f"{item}/{self.mapping_property.name}")
+                    particles = molecular_graph.molecular_groups[molecule][item]
+                    particle_trajectories = (
+                        tf.gather(batch[batch_reference], particles)
+                        * mass_dictionary[item]
+                    )
+                    molecule_trajectory += tf.reduce_sum(particle_trajectories, axis=0)
 
-        Returns
-        -------
-        Updates the database.
-        """
-        for molecule_name in self.molecules:
-            species = self.reference_molecules[molecule_name]["species"]
-            mass_factor = self._prepare_mass_array(species)
-            data_structure = self._prepare_database_entry(
-                molecule_name, len(self.adjacency_graphs[molecule_name]["molecules"])
-            )
-            path_list = [join_path(s, "Unwrapped_Positions") for s in species]
-            self._prepare_monitors(data_path=path_list)
-            scaling_factor = self.reference_molecules[molecule_name]["mass"]
-            molecules = self.experiment.molecules
-            molecules[molecule_name] = {}
-            molecules[molecule_name]["n_particles"] = len(
-                self.adjacency_graphs[molecule_name]["molecules"]
-            )
-            molecules[molecule_name]["indices"] = list(
-                range(molecules[molecule_name]["n_particles"])
+                # Compute the COM trajectory
+                trajectory[t, :, :] = molecule_trajectory
```

Comment:
`trajectory[t]` could be slightly faster than `trajectory[t, :, :]`

> 1.39 ms ± 38.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
vs.
> 1.37 ms ± 23.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

on a test array

Reply:
fb8ee66